### PR TITLE
Do not permit type parameters on builtin bounds.

### LINKED ISSUE #20302: We permit type parameters and so forth on builtin bounds
Example:

```
fn foo<T:Copy<U>, U>(x: T) {
}

fn main() {
}
```

I left a FIXME at the relevant part of the code.

### PR DESCRIPTION
**Pull request by ahmedcharles**
Fixes #20302.

r? @flaper87

### CODE REVIEW

File: src/test/compile-fail/typeck-builtin-bound-type-parameters.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn foo<T:Copy<U>, U>(x: T) {
+    //~^ ERROR: wrong number of type arguments: expected 0 found 1
```
Comment (oli-obk) [inline src/test/compile-fail/typeck-builtin-bound-type-parameters.rs]: is that span correct? shouldn't it be on the Copy<U> or just on the <U>?

Comment (FlaPer87) [inline src/test/compile-fail/typeck-builtin-bound-type-parameters.rs]: The `ERROR` match doesn't check on span positions. The `^` means the line above (L12) should raise this error. You can just remove the spaces in this `ERROR` line and it'll just work.

Comment (oli-obk) [inline src/test/compile-fail/typeck-builtin-bound-type-parameters.rs]: i don't know how i misread that :/

File: src/librustc_typeck/astconv.rs [hunk 2]
```diff
@@ -1731,8 +1731,14 @@ pub fn partition_bounds<'a>(tcx: &ty::ctxt,
...
                                                      &mut builtin_bounds) {
-                            // FIXME(#20302) -- we should check for things like Copy<T>
-                            continue; // success
+                            let segments = &b.trait_ref.path.segments;
+                            let parameters = &segments[segments.len() - 1].parameters;
+                            if parameters.is_empty() {
+                                continue; // success
+                            }
+                            span_fatal!(tcx.sess, b.trait_ref.path.span, E0316,
+                                        "wrong number of type arguments: expected 0 found {}",
```
Comment (FlaPer87) [inline src/librustc_typeck/astconv.rs]: I was wondering whether we could use a more specific error message since built-in bounds are special already:

`built-in bounds don't require type arguments, {} given`

**Comment by FlaPer87** [comment]
Other than that tiny nit about the error message, this looks good to me. @nikomatsakis thoughts on the error message?

**Comment by FlaPer87** [comment]
Also, great work and thanks for the patch. :cake:

File: src/test/compile-fail/typeck-builtin-bound-type-parameters.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn foo<T:Copy<U>, U>(x: T) {
```
Comment (FlaPer87) [inline src/test/compile-fail/typeck-builtin-bound-type-parameters.rs]: You may also want some other tests for things like:

```rust
trait Trait: Copy<Send> {}

struct MyStruct<T: Copy<T>>;
```

**Comment by ahmedcharles** [comment]
Ok, changed the message and added the tests.

File: src/librustc_typeck/astconv.rs [hunk 4]
```diff
@@ -1731,8 +1731,14 @@ pub fn partition_bounds<'a>(tcx: &ty::ctxt,
...
                                                      trait_did,
                                                      &mut builtin_bounds) {
-                            // FIXME(#20302) -- we should check for things like Copy<T>
-                            continue; // success
+                            let segments = &b.trait_ref.path.segments;
+                            let parameters = &segments[segments.len() - 1].parameters;
+                            if parameters.is_empty() {
+                                continue; // success
+                            }
+                            span_fatal!(tcx.sess, b.trait_ref.path.span, E0316,
```
Comment (nikomatsakis) [inline src/librustc_typeck/astconv.rs]: this should be `span_err`, not `span_fatal`

File: src/librustc_typeck/astconv.rs [hunk 5]
```diff
@@ -1731,8 +1731,14 @@ pub fn partition_bounds<'a>(tcx: &ty::ctxt,
...
+                                        "builtin bounds do not require type arguments, {} given",
+                                        parameters.types().len());
```
Comment (nikomatsakis) [inline src/librustc_typeck/astconv.rs]: this message doesn't seem right for something like `Copy<'a>` -- can you add a test for that case too? And adjust the message to be more generic? (Perhaps just "arguments" or "generic arguments")

**Comment by nikomatsakis** [comment]
Sorry for the delay! I forgot about this PR since I'm accustomed to just skimming what's assigned to me. :( Assigning to myself.

File: src/librustc_typeck/astconv.rs (hunk 4 above)
Comment (FlaPer87) [inline src/librustc_typeck/astconv.rs]: mmh, do you think we should make the other `span_fatal!` errors too? We currently consider this kind of errors as fatal errors for other types.

Comment (FlaPer87) [inline src/librustc_typeck/astconv.rs]: @ahmedcharles when changing this to `span_error` you should be able to merge the tests in a single test file.

Comment (nikomatsakis) [inline src/librustc_typeck/astconv.rs]: In general, every span_fatal is a bug imo, though sometimes there are good reasons. In this particular case, I think those other errors are holdovers and could be made span_err with some effort. (span_err can be trickier because you have to have some kind of intelligent fallback behavior...)

Comment (FlaPer87) [inline src/librustc_typeck/astconv.rs]: I fully agree with you. @ahmedcharles if you want to take a stab and make the other `span_fatal!` errors, it'd be great. Otherwise, I'll do it in a separate patch. :)

Comment (ahmedcharles) [inline src/librustc_typeck/astconv.rs]: I don't mind changing them, though I'd rather do it in a separate patch.

File: src/librustc_typeck/astconv.rs (hunk 5 above)
Comment (ahmedcharles) [inline src/librustc_typeck/astconv.rs]: I tried:

impl Copy<'a> for Foo {}

But it doesn't seem to work. Is there a better way to test lifetimes? Lifetimes is probably the most important feature of rust that I haven't wrapped my head around yet.

Comment (FlaPer87) [inline src/librustc_typeck/astconv.rs]: @ahmedcharles similar tests to the ones you have should be ok. For example:

`struct MyStruct<'a, T: Copy<'a>>;`

`fn foo<'a, T:Copy<'a, U>, U>(x: T) {}`

Trait implementations are checked correctly, the bug we're fixing is for trait bounds. That's why the example you tried is correctly throwing you an error already (undefined lifetime). If you try `impl<'a> Copy<'a> for Foo {}` it should throw an error saying that the number of the lifetimes specified is not the one expected.

**Comment by nikomatsakis** [comment]
@ahmedcharles ping, this looks almost ready to go, any chance you can address the various nits?

**Comment by ahmedcharles** [comment]
I've been busy this week and haven't had time to figure out the issues I've had with the tests that were suggested, some of them seem to produce multiple errors, instead of just the one I added. I need some more time to investigate.

**Comment by FlaPer87** [comment]
@ahmedcharles no worries at all. If you feel you won't have time, we can take the patch over and add the tests ourselves in a separate commit.

Also, I wonder if you read my last comment. I explained there what the issues with the tests you tried were. Let me know!

**Comment by pnkfelix** [comment]
@ahmedcharles hey there, do you think you'll have time to look at this pull request again?

If not, one of us can happily take over the patch for you; just let us know so we don't duplicate effort.

(I assume if we don't hear back, then we'll just take the patch over.)

**Comment by ahmedcharles** [comment]
I'll either have it fixed today or pass it off.

**Comment by ahmedcharles** [comment]
I'm not sure how to get the tests passing, but the current diff builds and mostly works as expected.

**Comment by FlaPer87** [comment]
I spoke with @ahmedcharles on IRC, I'll take this over and fix the tests preserving the existing commits.

**Comment by FlaPer87** [comment]
superseded by #22680
